util/spec_augment.py

```diff
@@ -0,0 +1,104 @@
+"""SpecAugment Implementation for Tensorflow.
+Related paper : https://arxiv.org/pdf/1904.08779.pdf
+In this paper, show summarized parameters by each open datasets in Tabel 1.
+-----------------------------------------
+Policy | W  | F  | m_F |  T  |  p  | m_T
+-----------------------------------------
+None   |  0 |  0 |  -  |  0  |  -  |  -
+-----------------------------------------
+LB     | 80 | 27 |  1  | 100 | 1.0 | 1
+-----------------------------------------
+LD     | 80 | 27 |  2  | 100 | 1.0 | 2
+-----------------------------------------
+SM     | 40 | 15 |  2  |  70 | 0.2 | 2
+-----------------------------------------
+SS     | 40 | 27 |  2  |  70 | 0.2 | 2
+-----------------------------------------
+LB : LibriSpeech basic
+LD : LibriSpeech double
+SM : Switchboard mild
+SS : Switchboard strong
+"""
+
+import librosa
+import librosa.display
+import tensorflow as tf
+from tensorflow.contrib.image import sparse_image_warp
+from tensorflow.python.framework import constant_op
+import numpy as np
+import random
+
+
+def spec_augment(mel_spectrogram, time_warping_para=80, frequency_masking_para=27,
+                 time_masking_para=100, frequency_mask_num=1, time_mask_num=1):
+    """Spec augmentation Calculation Function.
+    'SpecAugment' have 3 steps for audio data augmentation.
+    first step is time warping using Tensorflow's image_sparse_warp function.
+    Second step is frequency masking, last step is time masking.
+    # Arguments:
+      mel_spectrogram(numpy array): audio file path of you want to warping and masking.
+      time_warping_para(float): Augmentation parameter, "time warp parameter W".
+        If none, default = 80 for LibriSpeech.
+      frequency_masking_para(float): Augmentation parameter, "frequency mask parameter F"
+        If none, default = 100 for LibriSpeech.
+      time_masking_para(float): Augmentation parameter, "time mask parameter T"
+        If none, default = 27 for LibriSpeech.
+      frequency_mask_num(float): number of frequency masking lines, "m_F".
+        If none, default = 1 for LibriSpeech.
+      time_mask_num(float): number of time masking lines, "m_T".
+        If none, default = 1 for LibriSpeech.
+    # Returns
+      mel_spectrogram(numpy array): warped and masked mel spectrogram.
+    """
+    v = mel_spectrogram.shape[0]
+    tau = mel_spectrogram.shape[1]
```

Comment:
The shape of `mel_spectrogram` is `[batch_size=1, max_time=None, num_freqs_channels=257]`, so `v` and `tau` need to be adjusted properly.

Reply:
Also, the output of `contrib_audio.audio_spectrogram` is not a mel-scale spectrogram, so at a minimum, the argument name has to be updated. But the right thing to do here would be to convert the spectrogram to mel-scale in `feeding.py` before warping it, otherwise this is not doing the same thing as the SpecAugment paper. You can do that using `tf.signal.linear_to_mel_weight_matrix`. But note that `contrib_audio.mfcc` doesn't work if you pass in an already mel-scale spectrogram, so you would also need to replace it with `tf.signal.mfccs_from_log_mel_spectrograms`. The documentation for that function has some example code: https://www.tensorflow.org/api_docs/python/tf/signal/mfccs_from_log_mel_spectrograms.html

Reply:
I feel the following might be correct:
```
def get_log_mel_spectogram(pcm, sample_rate, frame_length, frame_step, fft_length):
  global v, tau
  
  stfts = tf.signal.stft(pcm, frame_length=frame_length, frame_step=frame_step, fft_length=fft_length)
  spectrograms = tf.abs(stfts)

  num_spectrogram_bins = stfts.shape[-1].value
  lower_edge_hertz, upper_edge_hertz, num_mel_bins = 80.0, 7600.0, 80
  linear_to_mel_weight_matrix = tf.signal.linear_to_mel_weight_matrix(
                                  num_mel_bins, \
                                  num_spectrogram_bins, \
                                  sample_rate, \
                                  lower_edge_hertz, \
                                  upper_edge_hertz
                                )
  mel_spectrograms = tf.tensordot( spectrograms, linear_to_mel_weight_matrix, 1)
  mel_spectrograms.set_shape(spectrograms.shape[:-1].concatenate(linear_to_mel_weight_matrix.shape[-1:]))

  log_mel_spectrograms = tf.log(mel_spectrograms + 1e-6)
  
  shape = tf.shape(log_mel_spectrograms)
  v = shape[1] 
  tau = shape[2]
  
  return log_mel_spectrograms
```

Reply:
That looks good but setting the v and tau globals is unnecessary (and probably a bug).

Reply:
Why this is necessary because of tf.contrib.image.sparse_image_warp can't take the image tensor with dynamic dimensions and our dimensions are [1, v, tau, 1], which depends on the audio duration.

Reply:
The value of `v` will be `None` (unknown/dynamic dimension) which will also break `sparse_image_warp`.

Reply:
If `sparse_image_wrap` requires static shapes/uses NumPy, then the only way I can see for using it in the data pipeline is by using `tf.py_func`, like this example: https://www.tensorflow.org/guide/datasets#applying_arbitrary_python_logic_with_tfpy_func

Reply:
Is anything here normalizing the spectrogram, so the spec_augment function can mask values just by setting them to zero (as it does)?  Otherwise it should probably substitute a mean value.

Reply:
By doing .cache, aren't you caching al the computed spectograms either in memory or in disk? by doing that, as far as I understand you are only applying the spec augmentation once instead of applying it every time you run an epoch . I think the cache step should be removed
``` python
    dataset = (tf.data.Dataset.from_generator(generate_values,
                                              output_types=(tf.string, (tf.int64, tf.int32, tf.int64)))
                              .map(entry_to_features, num_parallel_calls=tf.data.experimental.AUTOTUNE)
                              .cache(cache_path)
                              .window(batch_size, drop_remainder=True).flat_map(batch_fn)
                              .prefetch(num_gpus))
```

Reply:
You are both correct. The spectrograms need to be normalized and the `cache` call needs to be removed (or rather, conditioned on spec augment being used).

@cahuja1992: @Kailegh makes a good point here, maybe we should land this first without image warping.

Reply:
> Is anything here normalizing the spectrogram, so the spec_augment function can mask values just by setting them to zero (as it does)? Otherwise it should probably substitute a mean value.

I guess masking could be of two ways, one could be by replacing with zero, other could be with mean value. Anyways it is a noise.

Reply:
I started with a separate implementation of spec augment. 
https://github.com/cahuja1992/DeepSpeech/tree/master/util/spec_augment

I feel as of now we can do a session run for the calculation of the log_mel_spectogram then create a graph further for the augmentation.

Reply:
> If `sparse_image_wrap` requires static shapes/uses NumPy, then the only way I can see for using it in the data pipeline is by using `tf.py_func`, like this example: https://www.tensorflow.org/guide/datasets#applying_arbitrary_python_logic_with_tfpy_func

I tried OpenCV "warpPerspective", but the results of tensorflow sparse_image_warp are different from the openCV warpPerspective.